src/python/pants/backend/java/test/junit.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+from dataclasses import dataclass
+
+from pants.backend.java.compile.javac import CompiledClassfiles, CompileJavaSourceRequest
+from pants.backend.java.target_types import JavaTestsSources
+from pants.core.goals.test import TestFieldSet, TestResult
+from pants.engine.addresses import Addresses
+from pants.engine.fs import AddPrefix, Digest, MergeDigests
+from pants.engine.process import FallibleProcessResult, Process
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    CoarsenedTargets,
+    Targets,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.jvm.resolve.coursier_fetch import (
+    CoursierLockfileForTargetRequest,
+    CoursierResolvedLockfile,
+    MaterializedClasspath,
+    MaterializedClasspathRequest,
+    MavenRequirements,
+)
+from pants.jvm.resolve.coursier_setup import Coursier
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class JavaTestFieldSet(TestFieldSet):
+    required_fields = (JavaTestsSources,)
+
+    sources: JavaTestsSources
+
+
+@rule(desc="Run JUnit", level=LogLevel.DEBUG)
+async def run_junit_test(
+    coursier: Coursier,
+    field_set: JavaTestFieldSet,
+) -> TestResult:
+    transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest([field_set.address]))
+    coarsened_targets = await Get(
+        CoarsenedTargets, Addresses(t.address for t in transitive_targets.closure)
+    )
+    lockfile = await Get(
+        CoursierResolvedLockfile,
+        CoursierLockfileForTargetRequest(Targets(transitive_targets.closure)),
+    )
```

Comment:
cc @Eric-Arellano and @chrisjrn : Worth taking a look here to see how this aligns with your plans (see PR description as well). If we're able to find a semantic model for locked resolves that works across Python and the JVM, that would be a huge win in terms of consistency and teachability.